features/topology-map/plugins/org.opennms.features.topology.plugins.topo.graphml/src/main/java/org/opennms/features/topology/plugins/topo/graphml/GraphMLTopologyProvider.java

```diff
@@ -120,6 +107,26 @@ public GraphMLTopologyProvider(GraphMLGraph graph, GraphMLServiceAccessor servic
         }
     }
 
+    private void setNodeIdForVertex(GraphMLVertex vertex) {
+        if (Objects.isNull(vertex)) {
+            return;
+        }
+        if (Objects.isNull(vertex.getNodeID())) {
+            String foreignSource = (String) vertex.getProperties().get(GraphMLProperties.FOREIGN_SOURCE);
+            String foreignId = (String) vertex.getProperties().get(GraphMLProperties.FOREIGN_ID);
+            if (!Strings.isNullOrEmpty(foreignSource) && !Strings.isNullOrEmpty(foreignId)) {
+                OnmsNode onmsNode = m_serviceAccessor.getNodeDao().findByForeignId(foreignSource, foreignId);
+                if (onmsNode != null) {
+                    vertex.setNodeID(onmsNode.getId());
+                } else {
+                    LOG.warn("no node found for the given foreignSource ({}) and foreignId ({})", foreignSource, foreignId);
+                }
+            } else {
+                LOG.warn("nodeId is null, foreignSource ({}) and foreignId ({}) must be set both", foreignSource, foreignId);
```

Comment:
We should enrich the log message to indicate what we were trying to do:
`nodeId is null. In order to resolve the nodeId a foreignSource and foreignId must be set.`